src/daft-parquet/src/metadata.rs

```diff
@@ -12,11 +16,220 @@ fn metadata_len(buffer: &[u8], len: usize) -> i32 {
+                    logical_type,
+                    converted_type,
+                    fields,
+                }))
+            }
+        }
+    }
+}
+
+/// Applies field_ids to a ParquetType, returns `None` if the field ID is not found
+/// in the `field_id_mapping`, or if the field has no field ID.
+fn apply_field_ids_to_parquet_type(
+    parquet_type: ParquetType,
+    field_id_mapping: &BTreeMap<i32, Field>,
+) -> Option<ParquetType> {
+    let field_id = parquet_type.get_field_info().id;
+    if field_id
+        .and_then(|field_id| field_id_mapping.get(&field_id))
+        .is_some()
+    {
+        let mut pq_type_rewriter = ParquetTypeFieldIdRewriter { field_id_mapping };
+        let rewritten_pq_type = ParquetTypeWrapper(parquet_type)
+            .rewrite(&mut pq_type_rewriter)
+            .unwrap()
+            .0;
+        Some(rewritten_pq_type)
+    } else {
+        None
+    }
+}
+
+/// Applies field_ids to a parquet2 FileMetaData struct
+/// 1. Rename columns based on the `field_id_mapping`
+/// 2. Drop any columns that don't have a `field_id`, or that don't have a corresponding entry in the `field_id_mapping`
+fn apply_field_ids_to_parquet_file_metadata(
+    file_metadata: FileMetaData,
+    field_id_mapping: &BTreeMap<i32, Field>,
+) -> super::Result<FileMetaData> {
+    use parquet2::metadata::{ColumnChunkMetaData, SchemaDescriptor};
+
+    // Construct a new Parquet Schema after applying renaming using field_ids
+    let new_schema_descr = SchemaDescriptor::new(
+        file_metadata.schema_descr.name().to_string(),
+        file_metadata
+            .schema_descr
+            .fields()
+            .iter()
+            .filter_map(|pq_type| {
+                apply_field_ids_to_parquet_type(pq_type.clone(), field_id_mapping)
+            })
+            .collect(),
+    );
+
+    // Get a mapping of {field_id: ColumnDescriptor} to use in modifying the redundant ColumnDescriptors in our RowGroupMetadata
+    let field_id_to_column_descriptor_mapping = new_schema_descr
+        .columns()
+        .iter()
+        .filter_map(|col_descr| {
+            col_descr
+                .descriptor
+                .primitive_type
+                .field_info
+                .id
+                .map(|field_id| (field_id, col_descr))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    let new_row_groups = file_metadata
+        .row_groups
+        .iter()
+        .map(|rg| {
+            let new_columns = rg
+                .columns()
+                .iter()
+                .filter_map(|column| {
+                    let field_id = column.descriptor().descriptor.primitive_type.field_info.id;
+                    field_id
+                        .and_then(|field_id| field_id_to_column_descriptor_mapping.get(&field_id))
+                        .map(|col_descr| {
+                            ColumnChunkMetaData::new(
+                                column.column_chunk().clone(),
```

Comment:
avoid this clones since they may be expensive

Reply:
Ditto here - I can't access the private `.columns` in the RowGroupMetadata unless we make it public, since `.columns() -> &[ColumnChunkMetadata]`